dataflow/src/org/checkerframework/dataflow/cfg/DOTCFGVisualizer.java

```diff
@@ -65,25 +70,92 @@ public void init(Map<String, Object> args) {
         this.checkerName = (String) args.get("checkerName");
 
         this.generated = new HashMap<>();
+
+        this.sbStore = new StringBuilder();
     }
 
     /**
      * {@inheritDoc}
      */
-    public <A extends AbstractValue<A>, S extends Store<S>, T extends TransferFunction<A, S>>
-    /*@Nullable*/ Map<String, Object> visualize(ControlFlowGraph cfg, Block entry,
+    public /*@Nullable*/ Map<String, Object> visualize(ControlFlowGraph cfg, Block entry,
             /*@Nullable*/ Analysis<A, S, T> analysis) {
-        StringBuilder sb1 = new StringBuilder();
-        StringBuilder sb2 = new StringBuilder();
+
+        String dotgraph = generateDotGraph(cfg, entry, analysis);
+
+        String dotfilename = dotOutputFileName(cfg.underlyingAST);
+        // System.err.println("Output to DOT file: " + dotfilename);
+
+        try {
+            FileWriter fstream = new FileWriter(dotfilename);
+            BufferedWriter out = new BufferedWriter(fstream);
+            out.write(dotgraph);
+            out.close();
+        } catch (IOException e) {
+            ErrorReporter.errorAbort("Error creating dot file: " + dotfilename +
+                    "; ensure the path is valid", e);
+        }
+
+        Map<String, Object> res = new HashMap<>();
+        res.put("dotFileName", dotfilename);
+
+        return res;
+    }
+
+    /**
+     * generate the dot string write to dot file
+     * @param cfg
+     * @param entry
+     * @param analysis
+     * @return
+     */
+    protected String generateDotGraph(ControlFlowGraph cfg, Block entry,
+        /*@Nullable*/ Analysis<A, S, T> analysis) {
+        StringBuilder sbDigraph = new StringBuilder();
         Set<Block> visited = new HashSet<>();
-        Queue<Block> worklist = new LinkedList<>();
-        Block cur = entry;
-        visited.add(entry);
 
         // header
-        sb1.append("digraph {\n");
-        sb1.append("    node [shape=rectangle];\n\n");
+        sbDigraph.append("digraph {\n");
+
+        /*generateDotEdges must called BEFORE generateDotNodes,
```

Comment:
Clean up the English.
If the order of the method calls is important, I would suggest also reordering their appearance in the source code.